### PR TITLE
[LoopVectorize] Enable more early exit vectorisation tests

### PR DESCRIPTION
**Pull request by david-arm**
PR #112138 introduced initial support for dispatching to
multiple exit blocks via split middle blocks. This patch
fixes a few issues so that we can enable more tests to use
the new enable-early-exit-vectorization flag. Fixes are:

1. The code to bail out for any loop live-out values happens
too late. This is because collectUsersInExitBlocks ignores
induction variables, which get dealt with in fixupIVUsers.
I've moved the check much earlier in processLoop by looking
for outside users of loop-defined values.
2. We shouldn't yet be interleaving when vectorising loops
with uncountable early exits, since we've not added support
for this yet.
3. Similarly, we also shouldn't be creating vector epilogues.
4. Similarly, we shouldn't enable tail-folding.
5. The existing implementation doesn't yet support loops
that require scalar epilogues, although I plan to add that
as part of PR #88385.
6. The new split middle blocks weren't being added to the
parent loop.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 1]
```diff
@@ -5001,6 +5019,10 @@ LoopVectorizationCostModel::selectInterleaveCount(ElementCount VF,
   if (!Legal->isSafeForAnyVectorWidth())
     return 1;
 
+  // We don't attempt to perform interleaving for early exit loops.
+  if (Legal->hasUncountableEarlyExit())
+    return 1;
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:5024]: Could you add more details on why not? Is all that would be needed to deal with multiple parts in AnyOf & co?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 2]
```diff
@@ -10208,6 +10253,20 @@ bool LoopVectorizePass::processLoop(Loop *L) {
   if (UseInterleaved)
     IAI.analyzeInterleaving(useMaskedInterleavedAccesses(*TTI));
 
+  if (LVL.hasUncountableEarlyExit()) {
+    BasicBlock *LoopLatch = L->getLoopLatch();
+    if (IAI.requiresScalarEpilogue() ||
+        llvm::any_of(LVL.getCountableExitingBlocks(),
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:10259]:
```suggestion
       any_of(LVL.getCountableExitingBlocks(),
```

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 3]
```diff
@@ -10177,13 +10202,33 @@ bool LoopVectorizePass::processLoop(Loop *L) {
...
+    for (BasicBlock *BB : L->blocks()) {
+      for (Instruction &I : *BB) {
+        for (User *U : I.users()) {
+          Instruction *UI = cast<Instruction>(U);
+          if (!L->contains(UI)) {
+            reportVectorizationFailure(
+                "Auto-vectorization of loops with uncountable "
+                "early exit and live-outs is not yet supported",
+                "Auto-vectorization of loop with uncountable "
+                "early exit and live-outs is not yet supported",
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:10225]: Might be worth adding a variant that takes a single message instead of duplicating?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 4]
```diff
@@ -7813,11 +7835,14 @@ DenseMap<const SCEV *, Value *> LoopVectorizationPlanner::executePlan(
...
   // values from the additional bypass block.
   if (VectorizingEpilogue) {
+    assert(!ILV.Legal->hasUncountableEarlyExit() &&
+           "Epilogue vectorisation not yet supported with early exits");
     BasicBlock *BypassBlock = ILV.getAdditionalBypassBlock();
     for (VPRecipeBase &R : *ExitVPBB) {
       fixReductionScalarResumeWhenVectorizingEpilog(
           &R, State, State.CFG.VPBB2IRBB[ExitVPBB], BypassBlock);
     }
+
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7845]: unrelated?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 5]
```diff
@@ -4123,7 +4138,8 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
   // a bottom-test and a single exiting block. We'd have to handle the fact
   // that not every instruction executes on the last iteration.  This will
   // require a lane mask which varies through the vector loop body.  (TODO)
-  if (TheLoop->getExitingBlock() != TheLoop->getLoopLatch()) {
+  if (Legal->hasUncountableEarlyExit() ||
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:4141]: Is this needed? If there's an uncountable early exit, there won't be a single exiting block and the check below will be true (as there must be a latch)

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 6]
```diff
@@ -4753,7 +4769,9 @@ bool LoopVectorizationPlanner::isCandidateForEpilogueVectorization(
   // Epilogue vectorization code has not been auditted to ensure it handles
   // non-latch exits properly.  It may be fine, but it needs auditted and
   // tested.
-  if (OrigLoop->getExitingBlock() != OrigLoop->getLoopLatch())
+  // TODO: Add support for loops with an early exit.
+  if (Legal->hasUncountableEarlyExit() ||
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:4773]: Is this needed? If there's an uncountable early exit, there won't be a single exiting block and the check below will be true (as there must be a latch)

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 3 above)
Comment (david-arm) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Good suggestion! I've added one and I've started using it in this patch, but as a follow-on I'm happy to clean up any other cases in the vectoriser or legality code that take duplicated strings.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 5 above)
Comment (david-arm) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: You're right, the latch should always be non-null.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 6 above)
Comment (david-arm) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: You're right, the latch should always be non-null.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 7]
```diff
@@ -10177,13 +10201,32 @@ bool LoopVectorizePass::processLoop(Loop *L) {
...
+      return false;
+    }
+
+    // In addUsersInExitBlocks we already bail out if there is an outside use
+    // of a loop-defined variable, but it ignores induction variables which are
+    // handled by InnerLoopVectorizer::fixupIVUsers. We need to bail out if we
+    // encounter induction variables too otherwise fixupIVUsers will crash.
+    for (BasicBlock *BB : L->blocks()) {
+      for (Instruction &I : *BB) {
+        for (User *U : I.users()) {
```
[the remark this answers is not in the document]
Comment (david-arm) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Good suggestion. Done!

[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: I was a bit surprised to see those changes were needed given the existing checks in VPlan. I had a look to see why we were missing cases. Looks like we weren't properly handling exit phis with multiple operands properly, which should be fixable https://github.com/llvm/llvm-project/pull/120260

[the remark this answers is not in the document]
Comment (david-arm) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: This is an issue I believe I described in a previous vectoriser call and at the time I mentioned I had a fix for this already. I just haven't had chance to land the fix because it depended upon this patch landing first. In the spirit of landing small, incremental patches I decided it was best to run all the vectoriser tests with the new flag to get the code defended, then follow on with incremental fixes and additional functionality. This patch holds up several other patches and I would prefer not to make this patch dependent on #120260.

[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Do you have any concerns regarding https://github.com/llvm/llvm-project/pull/120260 or will the future patches rely on the IR based checks?

It is also possible to share multiple stacked PR to give a preview of what changes are in the pipeline (one option is to have a PR just include all commits or do stacked PRs via user branches https://discourse.llvm.org/t/update-on-github-pull-requests/71540/146?u=fhahn)

**Review by paulwalker-arm — APPROVED** [review]
This works for me.  Without looking deeper I cannot confirm if #120260 is valid but I figure there's no harm in landing all the fixes (and tests) in this PR and then #120260 can remove the one it improves upon when ready.

**Comment by fhahn** [comment]
> This works for me. Without looking deeper I cannot confirm if #120260 is valid but I figure there's no harm in landing all the fixes (and tests) in this PR and then #120260 can remove the one it improves upon when ready.

Yeah sounds good to me, I guess we can remove the extra checks again as part of #120260

**Comment by david-arm** [comment]
Hi, looks like a merge issue as the tests were passing before I merged. I'll revert the patch and reapply.

**Comment by fhahn** [comment]
Looks like the test updates were not in sync with current `main`, pushed a fix 3e02038948abf60d5c9a600f0a08a2dd9223982a to avoid the test failure

**Comment by david-arm** [comment]
> Looks like the test updates were not in sync with current `main`, pushed a fix [3e02038](https://github.com/llvm/llvm-project/commit/3e02038948abf60d5c9a600f0a08a2dd9223982a) to avoid the test failure

That's fast @fhahn - thank you! I was just about to look into it. I do normally rebase downstream and run `make check-all`, but for some reason I forgot today. Does rebasing the patch in github trigger new runs though? I don't entirely trust it to do the right thing.

**Comment by david-arm** [comment]
> @david-arm just noticed it failing locally. Pressing the `Update Branch` button should merge in changes from current `main` and then trigger a new run of the pre-commit checks

Well anyway, apologies for my silly mistake. I seem to remember in the past there were times I rebased and it didn't trigger a new run, but maybe I was doing something wrong. Even so, I "normally" prefer to be paranoid and run "make check-all" downstream, except for today when I didn't. :)
